File: apps/api/app/schemas/inventory.py
from datetime import datetime

from pydantic import BaseModel

from ._base import ORMModel


class InventoryLevelRead(ORMModel):
    id: str
    tenant_id: str
    store_id: str
    product_id: str
    on_hand: float
    safety_stock: float
    reserved: float
    updated_at: datetime
    store_name: str | None = None
    product_name: str | None = None
    product_sku: str | None = None


class InventoryLevelUpdate(BaseModel):
    safety_stock: float | None = None


class MovementCreate(BaseModel):
    """``store_id`` is optional; defaults to the JWT's store. ``user_id`` and
    ``terminal_id`` are likewise derived from the session."""

    id: str
    store_id: str | None = None
    product_id: str
    qty_delta: float
    reason: str
    ref_type: str | None = None
    ref_id: str | None = None
    terminal_id: str | None = None
    user_id: str | None = None
    note: str | None = None
    client_created_at: datetime


class MovementRead(ORMModel):
    id: str
    tenant_id: str
    store_id: str
    product_id: str
    qty_delta: float
    reason: str
    ref_type: str | None
    ref_id: str | None
    created_at: datetime


class TransferLineIn(BaseModel):
    id: str
    product_id: str
    qty: float
    received_qty: float | None = None


class TransferCreate(BaseModel):
    id: str
    from_store_id: str
    to_store_id: str
    status: str = "draft"
    note: str | None = None
    lines: list[TransferLineIn]


class TransferUpdate(BaseModel):
    status: str


class TransferRead(ORMModel):
    id: str
    tenant_id: str
    from_store_id: str
    to_store_id: str
    status: str
    dispatched_at: datetime | None
    received_at: datetime | None
    note: str | None
    created_at: datetime


class StocktakeLineIn(BaseModel):
    id: str
    product_id: str
    expected_qty: float
    actual_qty: float


class StocktakeCreate(BaseModel):
    id: str
    store_id: str | None = None
    note: str | None = None
    lines: list[StocktakeLineIn]


class StocktakeRead(ORMModel):
    id: str
    tenant_id: str
    store_id: str
    completed_at: datetime | None
    note: str | None
    created_at: datetime


class InventoryAdjustCreate(BaseModel):
    store_id: str
    product_id: str
    mode: str = "delta"  # delta | set
    qty: float
    note: str | None = None
    reason: str = "adjustment"  # adjustment | initial

File: apps/admin/src/api/inventory.ts
import client from './client'
import type { InventoryLevelRead, TransferRead, StocktakeRead } from '@/types'

export function listInventoryLevels(params?: { store_id?: string }) {
  return client.get<InventoryLevelRead[]>('/inventory/levels', { params })
}

export function updateInventoryLevel(id: string, data: { safety_stock?: number }) {
  return client.patch(`/inventory/levels/${id}`, data)
}

export function createTransfer(data: {
  id: string; from_store_id: string; to_store_id: string; note?: string
  lines: { id: string; product_id: string; qty: number }[]
}) {
  return client.post<TransferRead>('/inventory/transfers', data)
}

export function updateTransfer(id: string, data: { status: string }) {
  return client.patch<TransferRead>(`/inventory/transfers/${id}`, data)
}

export function createStocktake(data: {
  id: string; store_id: string; note?: string
  lines: { id: string; product_id: string; expected_qty: number; actual_qty: number }[]
}) {
  return client.post<StocktakeRead>('/inventory/stocktakes', data)
}

export function adjustInventory(data: {
  store_id: string
  product_id: string
  mode: 'delta' | 'set'
  qty: number
  note?: string
  reason?: 'adjustment' | 'initial'
}) {
  return client.post<InventoryLevelRead>('/inventory/adjust', data)
}
